OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/InAppMessageClickResult.kt

```diff
@@ -1,23 +1,138 @@
+
+    /**
+     * An optional URL that opens when the action takes place
+     */
+    override val url: String?
+
+    /**
+     * UUID for the page in an IAM Carousel
+     */
+    val pageId: String?
+
+    /**
+     * Outcome for action
+     */
+    val outcomes: MutableList<InAppMessageOutcome> = mutableListOf()
+
+    /**
+     * Prompts for action
+     */
+    val prompts: MutableList<InAppMessagePrompt> = mutableListOf()
+
+    /**
+     * Tags for action
+     */
+    var tags: InAppMessageTag? = null
+
+    /**
+     * Determines if this was the first action taken on the in app message
+     */
+    var isFirstClick = false
+
+    /**
+     * Determines if tapping on the element should close the In-App Message.
+     */
+    override val closingMessage: Boolean
+
+    init {
+        clickId = json.optString(ID, null)
+        actionId = json.optString(NAME, null)
+        url = json.optString(URL, null)
+        pageId = json.optString(PAGE_ID, null)
+
+        urlTarget = InAppMessageActionUrlType.fromString(json.optString(URL_TARGET, null))
+        if (urlTarget == null) {
+            urlTarget = InAppMessageActionUrlType.IN_APP_WEBVIEW
+        }
+
+        closingMessage = json.optBoolean(CLOSE, true)
+        if (json.has(OUTCOMES)) parseOutcomes(json)
+        if (json.has(TAGS)) tags = InAppMessageTag(json.getJSONObject(TAGS))
+        if (json.has(PROMPTS)) parsePrompts(json, promptFactory)
+    }
+
+    @Throws(JSONException::class)
+    private fun parseOutcomes(json: JSONObject) {
+        val outcomesJsonArray = json.getJSONArray(OUTCOMES)
+        for (i in 0 until outcomesJsonArray.length()) {
+            outcomes.add(InAppMessageOutcome((outcomesJsonArray[i] as JSONObject)))
+        }
+    }
+
+    @Throws(JSONException::class)
+    private fun parsePrompts(json: JSONObject, promptFactory: IInAppMessagePromptFactory) {
+        val promptsJsonArray = json.getJSONArray(PROMPTS)
+        for (i in 0 until promptsJsonArray.length()) {
+            val promptType = promptsJsonArray.getString(i)
+            val prompt = promptFactory.createPrompt(promptType)
+            if (prompt != null) {
+                prompts.add(prompt)
+            }
+        }
+    }
 
     fun toJSONObject(): JSONObject {
-        return JSONObject()
-            .put("message", _message.toJSONObject())
-            .put("action", _action.toJSONObject())
+        val mainObj = JSONObject()
+        try {
+            mainObj.put(CLICK_NAME, actionId)
+            mainObj.put(CLICK_URL, url)
+            mainObj.put(FIRST_CLICK, isFirstClick)
+            mainObj.put(CLOSES_MESSAGE, closingMessage)
+            val outcomesJson = JSONArray()
+            for (outcome in outcomes) outcomesJson.put(outcome.toJSONObject())
+            mainObj.put(OUTCOMES, outcomesJson)
+            if (tags != null) {
+                mainObj.put(TAGS, tags!!.toJSONObject())
+            }
+            // Omitted for now until necessary
```

Comment:
[Our docs](https://documentation.onesignal.com/docs/iam-sdk-methods#osinappmessageaction-class) specify this is mainly for debugging. When would this become necessary such that it should be uncommented?

Reply:
We should put this back as we want to expose this url_target to wrapper SDKs.